### PR TITLE
make @boolToInt always return a u1 

### LINKED ISSUE #7950: make @boolToInt always return a u1 even when the bool is comptime known
A core design pillar of the language is that code that works at runtime should be able to work at compile time (barring significant metaprogramming or weird pointer shenanigans).  This rule is broken here:
```zig
const signed_bool = @bitCast(i1, @boolToInt(value));
```
This code will compile with a runtime known `value` but not with a comptime known `value`.  The workaround is to use `@bitCast(i1, @as(u1, @boolToInt(value)))`, but this feels like a "trick" to make the code safe for use at comptime.  The fewer of those we have, the better.

Instead, `@boolToInt` should return a comptime known u1 when the parameter is comptime known, to be consistent.

### PR DESCRIPTION
**Pull request by tisonkun**
This closes #7950.

cc @andrewrk @mlugg in case this patch helps.

### CODE REVIEW

File: lib/std/math/ilogb.zig [hunk 1]
```diff
@@ -48,7 +48,7 @@ fn ilogbX(comptime T: type, x: T) i32 {
         }
 
         // offset sign bit, exponent bits, and integer bit (if present) + bias
-        const offset = 1 + exponentBits + @boolToInt(T == f80) - exponentBias;
+        const offset = 1 + exponentBits + @intCast(comptime_int, @boolToInt(T == f80)) - exponentBias;
```
Comment (Vexu) [inline lib/std/math/ilogb.zig:51]:
```suggestion
        const offset = 1 + exponentBits + @as(comptime_int, @boolToInt(T == f80)) - exponentBias;
```

File: src/Sema.zig [hunk 2]
```diff
@@ -18410,8 +18410,8 @@ fn zirBoolToInt(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError!A
     const operand = try sema.resolveInst(inst_data.operand);
     if (try sema.resolveMaybeUndefVal(operand)) |val| {
         if (val.isUndef()) return sema.addConstUndef(Type.initTag(.u1));
-        const bool_ints = [2]Air.Inst.Ref{ .zero, .one };
-        return bool_ints[@boolToInt(val.toBool())];
+        if (val.toBool()) return sema.addConstant(Type.initTag(.u1), Value.initTag(.one));
+        return sema.addConstant(Type.initTag(.u1), Value.initTag(.zero));
```
Comment (Vexu) [inline src/Sema.zig:18414]:
```suggestion
        if (val.toBool()) return sema.addConstant(Type.u1, Value.one);
        return sema.addConstant(Type.u1, Value.zero);
```

**Review by Vexu — COMMENTED** [review]
Small style nits.

File: src/Sema.zig (hunk 2 above)
Comment (Vexu) [inline src/Sema.zig:18414]: Github won't let me commit that suggestion, trying again
```suggestion
        if (val.toBool()) return sema.addConstant(Type.u1, Value.one);
        return sema.addConstant(Type.u1, Value.zero);
```

**Comment by tisonkun** [comment]
There seems some more type inconsistent. I'm handling it:

```
zig-local-cache/o/544ff04ac1c968334ce79488ef10cbe8/source.zig:1137:40: error: expected signed integer type, found 'u1'
zig-local-cache/o/13471bdccffe1aa306a0cac07212bc34/source.zig:1128:39: error: expected signed integer type, found 'u1'
zig-local-cache/o/13471bdccffe1aa306a0cac07212bc34/source.zig:1131:49: error: expected signed integer type, found 'u1'
/Users/runner/work/zig/zig/lib/std/math.zig:1687:57: error: overflow of integer type 'u1' with value '-1'
/Users/runner/work/zig/zig/lib/std/math.zig:1770:41: note: called from here
```

@Vexu could you avoid force push before we are pending for merging? Otherwise it's hard for me to fetch your changes.

**Comment by tisonkun** [comment]
```
/Users/m1/actions-runner/_work/zig/zig/lib/std/math.zig:1687:46: error: type 'i1' cannot represent integer value '1'
```

My bad, lol.

**Comment by Vexu** [comment]
There are plenty of other error you might want to start with if you skip the 3000 lines of noise.

**Comment by tisonkun** [comment]
Fix the test result issue. For translate C issue, locally it logs:

```
/Users/chenzili/.cache/zig/o/6d01653752a9036d704a3412c0963c2b/source.zig:755:40: error: expected signed integer type, found 'u1'
    if (foo(@bitCast(u8, @truncate(i8, @boolToInt(@as(c_int, 1) == @as(c_int, 2))))) != 0) {
                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
referenced by:
    comptime_0: /Users/chenzili/Brittani/zig/build/../lib/std/start.zig:63:50
    remaining reference traces hidden; use '-freference-trace' to see all reference traces

/Users/chenzili/.cache/zig/o/5d933b79a7f095bdc711fe11f6000956/source.zig:746:39: error: expected signed integer type, found 'u1'
    return @bitCast(u8, @truncate(i8, @boolToInt(false or true)));
                                      ^~~~~~~~~~~~~~~~~~~~~~~~~
referenced by:
    main: /Users/chenzili/.cache/zig/o/5d933b79a7f095bdc711fe11f6000956/source.zig:777:11
    comptime_0: /Users/chenzili/Brittani/zig/build/../lib/std/start.zig:63:50
    remaining reference traces hidden; use '-freference-trace' to see all reference traces

/Users/chenzili/.cache/zig/o/5d933b79a7f095bdc711fe11f6000956/source.zig:749:49: error: expected signed integer type, found 'u1'
    return @bitCast(c_short, @truncate(c_short, @boolToInt(@as(c_int, 0) < @as(c_int, 1))));
                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

May I ask where this translate logic stay?

**Comment by Vexu** [comment]
That is generated with `src/translate_c.zig`, it's a bit more complex so I can handle it for you if needed.

**Comment by tisonkun** [comment]
@Vexu Thanks for your information! I'll appreciate it if you can help in updating the related logics. I took a look at `translate_c.zig` and it seems some debug tricks are required to understanding how `@bitCast(u8, @truncate(i8` occurs. Manually, replace it with `@as(u8, ...` should work.

**Comment by tisonkun** [comment]
It seems now all aarch64 tests passed but x86_64 ones failed. How can our changes be related to arch?

Related logs:

```
2023-05-15T16:54:25.7209903Z [16/21] Running zig1.wasm to produce /home/ci/actions-runner5/_work/zig/zig/build-release/zig2.c
2023-05-15T16:54:27.1853797Z [17/21] Building C object CMakeFiles/zig2.dir/compiler_rt.c.o
2023-05-15T16:56:00.2799660Z [18/21] Building C object CMakeFiles/zig2.dir/zig2.c.o
2023-05-15T16:56:00.2800788Z zig2.c:346451:6: warning: incompatible pointer types assigning to 'const uint8_t (*)[14]' (aka 'const unsigned char (*)[14]') from 'const uint8_t *' (aka 'const unsigned char *') [-Wincompatible-pointer-types]
2023-05-15T16:56:00.2801468Z  t53 = (anon__42218_51){(uint8_t const *)((uint8_t const *)&objcopy_ElfFile_28true_29_parse__anon_42242__42242), (uintptr_t)14ul}.ptr;
2023-05-15T16:56:00.2802832Z      ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2023-05-15T16:56:00.2803733Z zig2.c:347145:8: warning: incompatible pointer types assigning to 'uint8_t (*)[4]' (aka 'unsigned char (*)[4]') from 'uint8_t *' (aka 'unsigned char *') [-Wincompatible-pointer-types]
2023-05-15T16:56:00.2804229Z   t124 = t45.ptr;
2023-05-15T16:56:00.2804459Z        ^ ~~~~~~~
2023-05-15T16:56:00.2805236Z zig2.c:348351:6: warning: incompatible pointer types assigning to 'const uint8_t (*)[14]' (aka 'const unsigned char (*)[14]') from 'const uint8_t *' (aka 'const unsigned char *') [-Wincompatible-pointer-types]
2023-05-15T16:56:00.2805890Z  t53 = (anon__42328_51){(uint8_t const *)((uint8_t const *)&objcopy_ElfFile_28true_29_parse__anon_42242__42242), (uintptr_t)14ul}.ptr;
2023-05-15T16:56:00.2806313Z      ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2023-05-15T16:56:00.2807313Z zig2.c:349054:8: warning: incompatible pointer types assigning to 'uint8_t (*)[4]' (aka 'unsigned char (*)[4]') from 'uint8_t *' (aka 'unsigned char *') [-Wincompatible-pointer-types]
2023-05-15T16:56:00.2807758Z   t123 = t45.ptr;
2023-05-15T16:56:00.2807946Z        ^ ~~~~~~~
2023-05-15T16:56:00.2808141Z 4 warnings generated.
2023-05-15T16:56:12.5592729Z [19/21] Linking CXX executable zig2
2023-05-15T17:30:05.8667933Z [20/21] Building stage3
2023-05-15T17:30:05.8861337Z [20/21] Install the project...
2023-05-15T17:30:05.9071517Z -- Install configuration: "Release"
2023-05-15T17:30:07.4168309Z + echo Looking for non-conforming code formatting...
2023-05-15T17:30:07.4169112Z Looking for non-conforming code formatting...
2023-05-15T17:30:07.4173688Z + stage3-release/bin/zig fmt --check .. --exclude ../test/cases/ --exclude ../build-debug --exclude ../build-release
2023-05-15T17:30:09.9443785Z + stage3-release/bin/zig build -Dtarget=arm-linux-musleabihf
2023-05-15T17:42:01.4598028Z + pwd
2023-05-15T17:42:01.4602291Z + stage3-release/bin/zig build test docs --maxrss 21000000000 -fqemu -fwasmtime -Dstatic-llvm -Dtarget=native-native-musl --search-prefix /home/ci/deps/zig+llvm+lld+clang-x86_64-linux-musl-0.11.0-dev.1869+df4cfc2ec --zig-lib-dir /home/ci/actions-runner5/_work/zig/zig/build-release/../lib
2023-05-15T17:46:49.2837231Z zig test test-c-abi-arm-linux-musleabihf-ReleaseFast ReleaseFast arm-linux-musleabihf: error: warning(link): unexpected LLD stderr:
2023-05-15T17:46:49.2839561Z ld.lld: warning: Linking two modules of different target triples: '/home/ci/actions-runner5/_work/zig/zig/build-release/zig-local-cache/o/fb7621e59f8626159be78416dc8c85b2/test-c-abi-arm-linux-musleabihf-ReleaseFast.o' is 'arm-unknown-linux-musleabihf' whereas 'ld-temp.o' is 'armv6kz-unknown-linux-musleabihf'
2023-05-15T17:46:49.2983730Z 
2023-05-15T17:46:49.2984194Z 
2023-05-15T17:46:49.2984553Z 
2023-05-15T17:48:28.4640985Z run test behavior-x86_64-linux-none-Debug-selfhosted: error: expected Illegal instruction at address 0x80b28c5
2023-05-15T17:48:28.4642533Z run test behavior-x86_64-linux-none-Debug-selfhosted: error: while executing test 'test.cast bool to int', the following command terminated with signal 4 (expected exited with code 0):
2023-05-15T17:48:28.4643617Z /home/ci/actions-runner5/_work/zig/zig/build-release/zig-local-cache/o/62f4419403852e5de38dd4ceeda3d227/test --listen=- 
2023-05-15T20:38:31.3948363Z Build Summary: 4399/4543 steps succeeded; 141 skipped; 1 failed; 143069/149266 tests passed; 6197 skipped (disable with -fno-summary)
2023-05-15T20:38:31.3948854Z test transitive failure
...
error: the following build command failed with exit code 1:
/home/ci/actions-runner5/_work/zig/zig/build-release/zig-local-cache/o/865459311d79db599ea2170f025eac9f/build /home/ci/actions-runner5/_work/zig/zig/build-release/stage3-release/bin/zig /home/ci/actions-runner5/_work/zig/zig /home/ci/actions-runner5/_work/zig/zig/build-release/zig-local-cache /home/ci/actions-runner5/_work/zig/zig/build-release/zig-global-cache test docs --maxrss 21000000000 -fqemu -fwasmtime -Dstatic-llvm -Dtarget=native-native-musl --search-prefix /home/ci/deps/zig+llvm+lld+clang-x86_64-linux-musl-0.11.0-dev.1869+df4cfc2ec --zig-lib-dir /home/ci/actions-runner5/_work/zig/zig/build-release/../lib
```

**Comment by jacobly0** [comment]
> How can our changes be related to arch?

Note that not all runners run the same tests, and that `aarch64-selfhosted` tests are currently disabled, but `x86_64-selfhosted` is expected to pass many of the behavior tests.

**Comment by tisonkun** [comment]
@jacobly0 Thank you! I've merged the latest master branch. Could you approve the CI workflow here?

**Comment by r00ster91** [comment]
The langref should be updated:
```diff
       {#header_open|@boolToInt#}
       <pre>{#syntax#}@boolToInt(value: bool) u1{#endsyntax#}</pre>
       <p>
       Converts {#syntax#}true{#endsyntax#} to {#syntax#}@as(u1, 1){#endsyntax#} and {#syntax#}false{#endsyntax#} to
                   {#syntax#}@as(u1, 0){#endsyntax#}.
       </p>
-      <p>
-      If the value is known at compile-time, the return type is {#syntax#}comptime_int{#endsyntax#}
-          instead of {#syntax#}u1{#endsyntax#}.
-      </p>
       {#header_close#}
```

Anyway, wouldn't it be better if we just remove `@boolToInt` from the language?

> I would be open to replacing boolToInt with bitCast, but I think I like bitCast a lot more than Andrew does. It would simplify the language a bit though.

> I think I have not expressed my love of `@bitCast` enough. It's great

**Comment by tisonkun** [comment]
@r00ster91 I'm updating the doc...

For drop `@boolToInt`, I check `@bitCast(u1, true)` work as expected. If it's accepted, then the work becomes clean up.

Since it has the similar effect of #7950 and #7950 has been accepted, I'll proceed this patch and see if dropping `@boolToInt` and replace its usage as `@bitCast` a good idea. Anyway, this patch align the manner of `@boolToInt` to `@bitCast(u1, boolvar)`

**Comment by tisonkun** [comment]
Docs updated.

**Comment by tisonkun** [comment]
ping @Vexu @jacobly0 @andrewrk can we make progress here?

**Comment by tisonkun** [comment]
> I'd like to hear Andrew's opinion on https://github.com/ziglang/zig/issues/15794 first.

@Vexu That sounds reasonable.

Although, as #7950 was accepted and #7950 doesn't conflict #15794, this pull request can even be a prerequisite of #15794 since it fixes all recipients of `@boolToInt` to expect a `u1` instead of `comptime_int`. It won't be regression which we should later "revert".

**Pull request merged**